### DISCUSSION TITLE
"DEVELOPER_ERROR" (Code 10) on Google Login -- have tried everything!

### DISCUSSION THREAD

**Discussion by asindhu** [supabase/supabase#27310] [Questions] [answered]
I can't get Google login working with Supabase for a react-native (Expo) app, at least in production. I have looked at the following:

1. [Supabase guide on Google login](https://supabase.com/docs/guides/auth/social-login/auth-google?queryGroups=platform&platform=react-native)
2. react-native-google-signin docs and [FAQ](https://react-native-google-signin.github.io/docs/troubleshooting)
3. [This YouTube video](https://www.youtube.com/watch?v=Bs3FyScXMZo) that's Flutter-specific but has applicable info in the first half or so
4. Dozens of StackOverflow questions

Here's the steps I've tried and the analysis so far.

I actually **have** been able to get it working on an Expo **development build**. However, it was **not** by following the Supabase docs (item 1 above), which says to create an **Android** OAuth token into Google Cloud with a SHA-1 key and input this into Supabase. It was by following some of the steps in the YT video (item 3 above) which tells you to configure a **web client** OAuth token into Google Cloud and then input the Client ID and Client Secret into Supabase. This actually works in a debug/development build.

However, when I create a production build with this config in the Google Play store, I get the DEVELOPER_ERROR (code 10) error.

Alternatively, I have tried following the Supabase instructions (item 1 above) to the exact letter and that doesn't work either, whether in development or release. These are the instructions that say to create an Android OAuth key and provide the SHA-1.

Supposedly, the DEVELOPER_ERROR comes when there is a SHA-1 mismatch. There are many many different instructions from different sources on how to get the right SHA-1 key for your Auth key, and I've tried all of them. I have tried:
- `./gradlew signingReport` and taking the ":app:signingReport" SHA-1 key. It's the same for me for debug & release.
- `eas credentials` and taking "SHA-1 Fingerprint" in the development, preview, or production options (again all 3 are the same SHA-1 key, although **different** from the key above from gradle)
-  `keytool -list -v -keystore ~/.android/debug.keystore` and taking the "SHA1 certificate fingerprint", again different from the other two
- Going to the Google Play developer console > App Signing > SHA-1 certificate fingerprint, again it's different

I have all four of these keys set up as Android OAuth keys in Google Cloud and added them to the Supabase authentication configuration. None of them work either in development or in a production build. I still just get the DEVELOPER_ERROR.

I am super stuck on this. I can't seem to find any information on how to move forward. The only way I can get sign-in to work is in a development build using the OAuth web client. I will say, one clear inconsistency I see is that from the react-native-google-signin FAQ (item 2 above) it says pretty clearly that if you are passing a "webClientId" into `GoogleSignin.configure()` that it must be of type **web client, not Android.**. Quoting it directly:

> If you're passing webClientId in configuration object to GoogleSignin.configure() make sure it's correct and that it is of type web (NOT Android!). 

Whereas, the Supabase instructions (item 1 above) tell you to create an Android OAuth key and then pass it in as the webClientId parameter. I would imagine that this inconsistency has something to do with the problem, but I'm not sure.

Any help would be greatly appreciated. I'm not sure I've seen evidence of anyone successfully getting this set up all way through and could really use some input on it. Thank you so much in advance!

**Comment by dshukertjr** [comment, marked as answer]
The error you are facing has nothing to do with Supabase but means that you have a misconfiguration in your Google sign-in setup. 

Code 10 means that you have the wrong SHA 1 certificate fingerprint set on the client ID. For the production build, make sure you create a Android Client ID with the production SHA 1 certificate fingerprint.

**Reply by asindhu** [reply]
Thanks so much for your response. I've tried creating Android keys with all the SHA-1 keys described above, including the production one that I grabbed from the Google Play developer console, and it is still not working...

**Reply by ifahadazam** [reply]
Did you find the solution?
I have same issue but in flutter.

**Reply by WellingtonDevBR** [reply]
any solution?

**Reply by rtainor** [reply]
I was facing this issue for like a week attempting to create a Google SignIn plugin for Unity Android, after many attemps, it seems that if you want to get an authentication token to send to your backend you must use a WEB CLIENT ID in the application not the Android Client Id one, but also seems that you must keep the Android application configured in Google Cloud Console with the sha1 and all that stuff

**Comment by 0xifis** [comment]
I am getting the same error. Followed the same debug pattern as OP. What strikes me as suspicious is that the
```js
GoogleSignin.configure({
    scopes: ['https://www.googleapis.com/auth/drive.readonly'],
    webClientId: "xxxx",
});
```
Where the plugin docs mention that it has to a client ID of android type but the property name seems to suggest it is web type.

**Reply by ChristianTrummer99** [reply]
Yes, same question. WHERE DO I PUT MY ANDROID CLIENT ID?? it doesn't work when I put it in the webClientId section.. which makes sense because it's an android client

I'm using Expo React Native with react-native-google-signin

The sha-1 stuff was simple enough - I've created a client ID for all unique SHA-1 fingerprints (debug.keystore, eas dev/prev/prod, etc.) but I don't know where to put the android client ID(s). I'm not even able to get an idToken from my `const userInfo = await GoogleSignin.signIn()` call, so supabase isn't even involved yet.

**Reply by kelvin0027** [reply]
Did you find a solution?

**Reply by rtainor** [reply]
You must use a WEB CLIENT ID but also keep the Android Client Id with your app data in Google Cloud Console or whatever you're using

**Comment by 0xifis** [comment]
https://github.com/orgs/supabase/discussions/27618 seems to have hit a similar problem

**Comment by AhmedAlsudairy** [comment]
Lets stick to this post 

I have same problem and I tried every possible  sha-1 keys   and the docs not clear

**Reply by dshukertjr** [reply]
The error is not related to Supabase-specific configuration and is just a general error from setting up Google sign-in. I would suggest going through how to setup Google sign-in, and make sure that you are able to perform Google sign-in on your app before jumping into any Supabase related settings.

**Reply by codebrain22** [reply]
I have the same issue in Flutter but only in prod. Both debug and release SHA keys are attached on my Firebase project. What's surprising is that it was giving 12500 error, now it's just says 10. I tried almost every possible solution I could find on the internet but I'm not winning.

**Comment by realfelixyu** [comment]
Any update on this?

**Reply by dshukertjr** [reply]
The error is not related to Supabase-specific configuration and is just a general error from setting up Google sign-in. I would suggest going through how to setup Google sign-in, and make sure that you are able to perform Google sign-in on your app before jumping into any Supabase related settings.

**Comment by ivanff** [comment]
I tried the same thing with all the sha1 that I could find, and there is nothing that does not fit

the documentation is terrible

for this reason, I used native script and there were no problems with authorization in Google

**Comment by sqadan** [comment]
Just wanted to share something crucial that solved this issue for me. If you're using a cloud build system like EAS (Expo Application Services), the SHA-1 fingerprint you see locally might NOT be the one actually used in your built app.

I tried everything suggested (checking local debug.keystore, running ./gradlew signingReport, etc.) but kept getting the DEVELOPER_ERROR. The breakthrough came when I realized that EAS uses its own signing process, so I needed to:

1. Download the actual built APK from EAS
2. Extract its SHA-1 using:
   ```
   keytool -printcert -jarfile your-app.apk
   ```
3. Use THAT SHA-1 in Google Cloud Console

The SHA-1 from the built APK was completely different from what I saw locally! After updating Google Cloud Console with this SHA-1, Google Sign-In started working immediately.

If you're using any cloud build service (EAS, Flutter, etc.), make sure to check the SHA-1 from your actual built APK, not just your local signing reports.

**Reply by shaoweipng** [reply]
@sqadan : Can you share here which clientID you used for the below code in your supabase code? Did you use an android OAuth client or a web OAuth client?

GoogleSignin.configure({
    scopes: ['https://www.googleapis.com/auth/drive.readonly'],
    webClientId: "xxxx",
});

I am having the same confusion as @0xifis

**Reply by ifahadazam** [reply]
Will it work in release mode while the app is in internal testing. Works fine in debug.

Cuz the recommend way is that, in production we should use sha1 key from app signing certificate from play console and use it in cloud project of google auth.

**Reply by ChristianTrummer99** [reply]
@shaoweipng Same question. I am tearing my hair out. There seems to be no mention of where/what to do with the Android Client ID. My assumption _was_ that I would not need a web client ID as I'm not building a web app. 

Super weird how the android use case seems to be ignored in the GoogleSignin  API docs.

**Comment by TasseDeCafe** [comment]
For anyone trying to build with expo locally (without EAS!) with this command:

`npx expo run:android --device`

Ignore "eas credentials" or any instructions telling you to go in your home folder. When you run the build command, Expo will create an /android folder in the folder of the app. In /android/app, there is a debug.keystore file that is auto-generated. This file is the one to be used to generate the SHA-1 fingerprint for local builds with this command (when ran from apps/native):

`keytool -list -v -keystore android/app/debug.keystore -alias androiddebugkey -storepass android -keypass android`

This will generate the correct fingeprint.

**Reply by khalilmissaoui** [reply]
this worked for me , generate creeds with EAS were not efficient . if that is not working with u , try this one .

**Reply by monilraycha** [reply]
this worked for me also

**Comment by shaoweipng** [comment]
@asindhu did you figure out the solution? I am facing the same problem as you now.

**Comment by hambergerpls** [comment]
![image](https://github.com/user-attachments/assets/6ccfc7f3-ffb6-43b9-9f51-040cfae61fe5)

I'm not sure if this is the same error that we are discussing, but I'm facing similar issue when trying to set up native Google Sign In.

I narrowed down the cause by not running my local Supabase and try to log in and found that it throws an error even without making any requests to Supabase.

**The cause for the issue was that I input the wrong package name.**

I created my flutter app using the following command
```sh
flutter create . --org com.myapp --platforms=android,ios,web
```

So when setting up the OAuth client I put the following package name
![image](https://github.com/user-attachments/assets/c5062cdd-c075-4ab1-a5e9-2df200b2fcd0)

When in reality, my `android/app/build.gradle` file contains the following:
![image](https://github.com/user-attachments/assets/f9c45885-7576-4f99-9587-6696d9e0850e)

The fix was changing the package name to match the applicationId in gradle:
![image](https://github.com/user-attachments/assets/9fecc5d0-8ad5-4d8c-baa9-fcfb34f7a525)

For SHA-1 certificate fingerprint, just use the `keytool` command. In my case, my `debug.keystore` was located at my root project folder at `.android/debug.keystore` after running `flutter run`

My supabase google clientid uses the web client id created in Google Cloud console.

My Flutter version is `3.24.4` at the time of writing.

Hope this helps anyone who are still facing this issue!

**Comment by ifahadazam** [comment]
Anyone got working solution.
Similar problem struck me and have been debugging it for 2 day but can't find solution of it.

Framework: flutter
Backend : supabase 
Google auth Api from Google cloud.

1. Working fine in debug.
2. In play store internal testing it throws Error 10 despite using play signing Sha1 key into cloud project.
3. Using web client id in code of google sign in and supabase google auth config.

Its frustrating. Anyone got this?

**Comment by ASHWINI-GUPTA** [comment]
I’ve faced a similar issue, and to clarify - Supabase isn’t the root cause here. The actual problem is related to incorrect SHA-1 configuration in the Google Cloud Console.

When you run:
```bash
flutter run
```

You’ll get output like:

```log
Running Gradle task 'assembleDebug'...                             36.6s
✓ Built build/app/outputs/flutter-apk/app-debug.apk
```

To get the SHA-1 fingerprint of the debug certificate used to sign the APK, you can run:

```bash
keytool -printcert -jarfile build/app/outputs/flutter-apk/app-debug.apk
```
Sample output:

```log
Certificate #1:
Owner: C=US, O=Android, CN=Android Debug
Issuer: C=US, O=Android, CN=Android Debug
Serial number: 1
Valid from: Tue Oct 24 16:27:57 IST 2023 until: Thu Oct 16 16:27:57 IST 2053
Certificate fingerprints:
         SHA1: <paste-this-in-Google-Console>
         SHA256: ...
Signature algorithm name: SHA1withRSA (weak)
Subject Public Key Algorithm: 2048-bit RSA key
Version: 1
```

Now,
Copy the SHA1 fingerprint from the above and add it in the Google Console under:
- APIs & Services → Credentials → OAuth 2.0 Client IDs
- Add a new OAuth Client ID for your debug environment and paste the SHA-1 there.

> You can create multiple OAuth 2.0 Client IDs for different environments (e.g., debug, staging, release), each with its own SHA-1.

Hope this helps someone running into similar issues!

**Comment by AVu120** [comment]
I had the same issue, local development build (built using eas build with development profile) had google login working but production build (built using eas build with production profile) did not (kept getting _com.google.android.gms.common.api.ApiException DEVELOPER_ERROR_ error). 

The solution for me was to:

1. Go into Google Play console > select my app
2. On left-hand-side click on 'Test and Release' >  Setup > App signing
3. Copy the SHA-1 certificate fingerprint of the App signing key certificate 
4. Go to Google Cloud Console > APIs and services > Credentials
5. Create another OAuth client ID with Application type = 'Android'
6. Fill in all the fields & paste the SHA-1 certificate fingerprint of the App signing key certificate from step 3 into the 'SHA-1 certificate fingerprint' field.
7. Click 'Create' button.
8. Test 'Log in with Google' again on production build after a few minutes or hours and it worked (for me at least).

**Reply by cjmling** [reply]
OMG this worked!! and why the hell `eas credentials` for production have different keys :S 

// For me i didn't have to wait, it worked instantly.

**Reply by tacshade** [reply]
It looks like they have moved this now to under App Integrity. I was having the exact same issue and could not for the life of me figure out why this did not work.

<img width="940" height="158" alt="image" src="https://github.com/user-attachments/assets/7d8c2fab-3bec-4f3e-8b4b-5c82b1529e6a" />

**Reply by simon-wexer** [reply]
You legend. This worked for me. 
hitting my head with this and couldn’t understand what was happening with Google sign in and my production build from EAS. 

Adding this SHA-1 key from here immediately solved my problem 🫶🫶🫶

**Reply by jsh9611** [reply]
Thank you so much! :)
I have the same issue in Flutter, I found a new SHA-1 in the Google Play Console. 
After adding it to the Google Cloud Console, everything started working properly.

**Reply by gustavokenzo1** [reply]
You are a legend! I had all the SHA-1s from eas credentials but none of them worked. The one within App Integrity was the only one that worked for production! Since I am using Firebase, it was even easier: go to project configuration and add the SHA-1 in Android, no need to go to google cloud console!

**Reply by MrMoovf** [reply]
THIS IS THE ANSWER!!

**Comment by next6leo** [comment]
Thank you @AVu120 your explanation help me alot to fix this developer error

**Comment by ChristianTrummer99** [comment]
This is what solved it for me. To anyone who is still having issues, specifically with Google Sign In on **Android with Expo React Native NOT using Firebase**:

Yes, the docs on supabase are misleading (you **do in fact** need an OAuth Web Client in addition to Android and iOS clients in GCP). 

1. Get your SHA fingerprints figured out. This is pretty easy to do; see other msgs^.  Paste them all down on a note so you have _all_ of them in one place and you can test them one-by-one in GCP. Note, for production, you'll end up with multiple Android Clients in GCP (each with a different SHA fingerprint, assuming you have different fingerprints for different build types, eg. production, development, preview).
2. **Here's where I was getting stuck:** Make sure you have created _both_ Android and Web OAuth clients in GCP. 
3. Paste the Web Client ID in the `.configure()` call as per the various docs (supabase, expo, RN Google Sign In). 
4. The Android Client ID does not need to be explicitly referenced because the package name/fingerprint are used under-the-hood to find the appropriate Android OAuth client in GCP. 

Why both? Grok explanation pasted below. TLDR; the Web Client is not only used for web apps. Android Apps need both a web client and android client for google sign in. Android Client "verifies the app's identity" and the Web Client enables the ID token issuance (which is what gets passed to supabase). 

When creating your web client, assuming you're not building a web app, you can ignore the JS origins section and put `http://localhost` in the redirects section (which works for testing). 

My prompt:
>the various docs imply that I am supposed to create an "Android" type Oauth client when using android/ios as the google sign in client. There is no way, that I can tell, to pass in android client ID to the GoogleSignIn methods, im building an Android/IOS mobile app, not web app. Do i still need a web client? if i have a web client, does that mean I don't need to create an android client in GCP?

Grok answer
>Yes, the docs are a bit confusing on this point, but based on how Google Sign-In works under the hood (even in a React Native context with this library), you do need to create an Android-type OAuth client in Google Cloud Platform (GCP) for your mobile app. This is primarily to register your app's package name and SHA-1 signing certificate fingerprint, which allows Google to verify that sign-in requests are coming from your legitimate Android app. Without it, the sign-in flow may fail entirely or be insecure.
>
>However, you're correct that there's no direct parameter in the @react-native-google-signin/google-signin library's methods (like configure() or signIn()) to pass in the Android client ID. That's because the Android client ID isn't used in your code—it's handled automatically by Google's servers during the authentication process, based on the credentials tied to your GCP project.
>
>You still need a web client ID (of type "Web application" in GCP), even though you're building a mobile app and not a web app. This is specifically required to enable fetching the idToken from the sign-in response, which is what Supabase (or any backend) needs for secure verification. The web client ID is passed to GoogleSignin.configure({ webClientId: 'your-web-client-id' }) in your code. Without it, you'll get user info but a null idToken, as you're seeing.
>
>Having a web client does not mean you can skip creating the Android client—they serve different purposes:
>
> - Android client: Verifies your app's identity (via package name and SHA-1). This is mandatory for Android sign-in to work at all.
> - Web client: Enables ID token issuance (for backend auth). This is mandatory if you need the idToken (which you do for Supabase).
>
>In practice, when you create the Android client in GCP, Google often auto-generates a web client for your project if one doesn't exist (it's listed under "Web client (auto created by Google Service)" in the credentials page). You can use that auto-generated one, or create a manual web client if needed. You'll also want an iOS client for your iOS build.

Note: GCP did not auto-generate a web client for me.
